File: 프로그래머스/lv3/김원래/여행경로.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <iostream>

using namespace std;

unordered_map<string, vector<string>> map;
unordered_map<string, vector<int>> visit;
int n;

bool backtrack(string& curPos, vector<string>& curRes) { // 처음에 ICN에 push된 curRes에서 시작한다.
    curRes.push_back(curPos);
    
    //종료조건
    if(curRes.size() == n) return true;
    
    vector<string>& _paths = map[curPos];
    vector<int>& _visit = visit[curPos];
    
    for(int cnt = 0 ; cnt < _paths.size() ; ++cnt){
        if(_visit[cnt] == 1) continue;
        
        _visit[cnt] = 1;
        
        if(backtrack(_paths[cnt], curRes) == true) return true;
        
        _visit[cnt] = 0;
        curRes.pop_back();
    }
    
    return false;
}


vector<string> solution(vector<vector<string>> tickets) {
    vector<string> answer;
    
    n = tickets.size() + 1;
    
    for(vector<string> ticket : tickets) {
        if(map.find(ticket[0]) == map.end()) {
            map.insert({ticket[0], vector<string>()});
            visit.insert({ticket[0], vector<int>()});
        }
        map[ticket[0]].push_back(ticket[1]);
        visit[ticket[0]].push_back(0);
    }
    
    for(auto iter = map.begin() ; iter != map.end() ; ++iter) {
        vector<string>& adj = (*iter).second;
        sort(adj.begin(), adj.end());
    }
    
    string start = "ICN";
    
    backtrack(start, answer);
    
    return answer;
}

File: 프로그래머스/lv3/이하은/여행경로.kt
import java.util.*

/**
주어진 항공권을 모두 이용한 여행경로 구하기 (DFS)
ICN 노드에서 출발 (노드 최대 1만개)
가능한 경로가 2개 이상 -> 알파벳 순서가 앞서는 경로

A - B
C - D 
노드 식별자가 숫자가 아니라 문자열이다. 
*/
class Solution {
    val graph = HashMap<String, ArrayList<String>>()
    
    fun solution(tickets: Array<Array<String>>): Array<String> {
        var answer = arrayOf<String>()
        
        // 그래프 생성
        for(ticket in tickets){
            val start = ticket[0]
            val end = ticket[1]
            graph.getOrPut(start) { arrayListOf() }.add(end)
        }
        
        // 알파벳 순서대로 경로 정렬
        for((key, value) in graph){
            value.sort()
        }
        
        // 그래프 확인
        for((start, destinations) in graph){
            println("$start -> $destinations")
        }
        
        return answer
    }
}
